test/opt/test_set_spec_const_default_value.cpp

```diff
@@ -0,0 +1,429 @@
+// Copyright (c) 2016 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "pass_fixture.h"
+
+namespace {
+using namespace spvtools;
+
+using SpecIdToValueStrMap =
+    opt::SetSpecConstantDefaultValuePass::SpecIdToValueStrMap;
+
+struct SetSpecConstantDefaultValueTestCase {
+  const char* code;
+  SpecIdToValueStrMap default_values;
+  const char* expected;
+};
+
+using SetSpecConstantDefaultValueParamTest =
+    PassTest<::testing::TestWithParam<SetSpecConstantDefaultValueTestCase>>;
+
+TEST_P(SetSpecConstantDefaultValueParamTest, TestCase) {
+  const auto& tc = GetParam();
+  SinglePassRunAndCheck<opt::SetSpecConstantDefaultValuePass>(
+      tc.code, tc.expected, /* skip_nop = */ false, tc.default_values);
+}
+
+INSTANTIATE_TEST_CASE_P(
+    ValidCases, SetSpecConstantDefaultValueParamTest,
+    ::testing::ValuesIn(std::vector<SetSpecConstantDefaultValueTestCase>{
+        // 0. Empty.
+        {"", SpecIdToValueStrMap{}, ""},
+        // 1. Empty with non-empty values to set.
+        {"", SpecIdToValueStrMap{{1, "100"}, {2, "200"}}, ""},
+        // 2. Bool type.
+        {
+            // code
+            "OpDecorate %1 SpecId 100\n"
+            "OpDecorate %2 SpecId 101\n"
+            "%bool = OpTypeBool\n"
+            "%1 = OpSpecConstantTrue %bool\n"
+            "%2 = OpSpecConstantFalse %bool\n",
+            // default values
+            SpecIdToValueStrMap{{100, "false"}, {101, "true"}},
+            // expected
+            "OpDecorate %1 SpecId 100\n"
+            "OpDecorate %2 SpecId 101\n"
+            "%bool = OpTypeBool\n"
+            "%1 = OpSpecConstantFalse %bool\n"
+            "%2 = OpSpecConstantTrue %bool\n",
+        },
+        // 3. 32-bit int type.
+        {
+            // code
+            "OpDecorate %1 SpecId 100\n"
+            "OpDecorate %2 SpecId 101\n"
+            "%int = OpTypeInt 32 1\n"
+            "%1 = OpSpecConstant %int 10\n"
+            "%2 = OpSpecConstant %int 11\n",
+            // default values
+            SpecIdToValueStrMap{{100, "2147483647"}, {101, "0xffffffff"}},
```

Comment:
I'd like to see a case for 32-bit int where the string value has a negative sign.

E.g.   "-42" should go through.  That makes sure you've asked to parse the right type.